core/src/main/scala/kafka/log/Log.scala

```diff
@@ -652,6 +653,12 @@ class Log(@volatile private var _dir: File,
     bytesTruncated
   }
 
+  def segmentRecovery(): LogSegment => Int = {
+    (segment: LogSegment) => {
+      recoverSegment(segment, None)
```

Comment:
Log.recoverSegment() updates the producerState. If we are just rebuilding the indexes, we don't need to change producerState. Perhaps we could introduce a new method LogSegment.rebuildIndex() that does the same index rebuilding logic as LogSegment.recovery(), but without touching the producerState and leaderEpoch. If we encounter CorruptRecordException or InvalidRecordException, we reset the recovery point and fail the broker.

Reply:
This sounds reasonable to me, what do I need to do to ensure the recovery point is reset and the broker is failed?

Reply:
You can see how the recovery point is set in Log.flush(). To fail a broker, we can probably do kafka.utils.Exit.exit(e.statusCode).

It would be useful if you could preserve the commit history for easier incremental reviews.

Reply:
This behavior doesn't seem intuitive to me, as it is expected that indexes are trimmed when written to disk, so if they're loaded untrimmed it is unlikely the next write to disk will fix this issue leaving the broker in a crash loop. This is definitely better than deleting data prematurely, but still doesn't seem great.

Reply:
Well, if we rebuild the index on sanity check failure, the size in memory will be correct and the index is usable. When the index is reloaded after broker restart, the sanity check will fail again and we will rebuild the index again. So, we won't be in a crash loop, but we  have the cost of repeatedly rebuilding the index on broker restart. This seems to be the same behavior pre KIP-263. We can log a warning so that people can notice this potential issue.

Reply:
Thanks for the clarification, I was under the impression `kafka.utils.Exit.exit(e.statusCode)` would force the broker to shutdown. I will get these changes added, thanks for helping me navigate through this.

Reply:
We will only call `Exit.exit()` if we encounter CorruptRecordException or InvalidRecordException during index rebuild since that indicates the log itself is corrupted.